Added tutorials to documentation (#173)
con/nwb2bids: docs/tutorials.rst

.. code-block:: text

    ephys_tutorial_dataset/some_sessions_bids_dataset/
    ├── dataset_description.json
    ├── participants.tsv
    ├── participants.json
    └── sub-subject+1/
        ├── sub-subject+1_sessions.tsv
        ├── sub-subject+1_sessions.json
        ├── ses-session+1/
        │   └── ecephys/
        │       ├── sub-subject+1_ses-session+1_ecephys.nwb
        │       ├── sub-subject+1_ses-session+1_channels.tsv
        │       ├── sub-subject+1_ses-session+1_channels.json
        │       ├── sub-subject+1_ses-session+1_electrodes.tsv
        │       ├── sub-subject+1_ses-session+1_electrodes.json
        │       ├── sub-subject+1_ses-session+1_probes.tsv
        │       └── sub-subject+1_ses-session+1_probes.json
        └── ses-session+2/
            └── ecephys/
                ├── sub-subject+1_ses-session+2_ecephys.nwb
                ├── sub-subject+1_ses-session+2_channels.tsv
                ├── sub-subject+1_ses-session+2_channels.json
                ├── sub-subject+1_ses-session+2_electrodes.tsv
                ├── sub-subject+1_ses-session+2_electrodes.json
                ├── sub-subject+1_ses-session+2_probes.tsv

>                 └── sub-subject+1_ses-session+2_probes.json

Author: Yes, a future PR will add `.json` 'bubbling' up to the top level here and the docs will then be updated accordingly

Reviewer: - #67

.. _tutorials:

Tutorials
=========

**nwb2bids** comes pre-packaged with some demonstrative tools to help showcase its functionality on various types of
NWB file contents. No additional requirements, knowledge, or experience are necessary to run these tutorials!

These tutorials demonstrates how to convert NWB file(s) into a BIDS directory structure. Each action allows for
alternative use cases based on the command line interface (CLI) or the Python library.

.. _tutorial-single-file:

Tutorial 1 - Converting a single file
-------------------------------------

In case you don't have any NWB files handy, or perhaps you've never worked with NWB files before, you can generate a
toy example through the following methods:

.. tabs::

    .. tab:: CLI

        .. code-block:: bash

            nwb2bids tutorial ephys file

    .. tab:: Python Library

        .. code-block:: bash

            import nwb2bids

>             nwb2bids.testing.generate_ephys_tutorial(mode="file")

Reviewer 2: without `output_directory` this fails.
```
>>> import nwb2bids
>>> nwb2bids.testing.generate_ephys_tutorial(mode="file")
Traceback (most recent call last):
  File "<python-input-7>", line 1, in <module>
    nwb2bids.testing.generate_ephys_tutorial(mode="file")
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/home/austin/devel/nwb2bids/venv/lib64/python3.13/site-packages/pydantic/_internal/_validate_call.py", line 39, in wrapper_function
    return wrapper(*args, **kwargs)
  File "/home/austin/devel/nwb2bids/venv/lib64/python3.13/site-packages/pydantic/_internal/_validate_call.py", line 136, in __call__
    res = self.__pydantic_validator__.validate_python(pydantic_core.ArgsKwargs(args, kwargs))
pydantic_core._pydantic_core.ValidationError: 1 validation error for generate_ephys_tutorial
output_directory
  Missing required keyword only argument [type=missing_keyword_only_argument, input_value=ArgsKwargs((), {'mode': 'file'}), input_type=ArgsKwargs]
    For further information visit https://errors.pydantic.dev/2.12/v/missing_keyword_only_argument
```

Also for tutorial 2, but seems like the fix is to make this arg actually optional.

                .. code-block:: bash

                    nwb2bids convert ~/.nwb2bids/tutorials/ephys_tutorial_file/ephys.nwb \
                        --bids-directory ~/.nwb2bids/tutorials/ephys_tutorial_file/bids_dataset \
                        --additional-metadata-file-path ~/.nwb2bids/tutorials/ephys_tutorial_file/metadata.json

            .. tab:: Windows

                .. code-block:: bash

                    nwb2bids convert ~/.nwb2bids/tutorials/ephys_tutorial_file/ephys.nwb ^
                        --bids-directory ~/.nwb2bids/tutorials/ephys_tutorial_file/bids_dataset ^
                        --additional-metadata-file-path ~/.nwb2bids/tutorials/ephys_tutorial_file/metadata.json

    .. tab:: Python Library

        .. code-block:: python

            import pathlib

            import nwb2bids

            ephys_tutorial_file_directory = pathlib.Path.home() / ".nwb2bids/tutorials/ephys_tutorial_file"
            nwb_paths = [ephys_tutorial_file_directory / "ephys.nwb"]
            bids_directory = ephys_tutorial_file_directory / "bids_dataset"
            additional_metadata_file_path = ephys_tutorial_file_directory / "metadata.json"

>             nwb2bids.convert_nwb_dataset(nwb_paths=nwb_paths, bids_directory=bids_directory)

Author: Derp

                .. code-block:: bash

                    nwb2bids convert ~/.nwb2bids/tutorials/ephys_tutorial_dataset/some_sessions \
                        --bids-directory ~/.nwb2bids/tutorials/ephys_tutorial_dataset/some_sessions_bids_dataset

            .. tab:: Windows

                .. code-block:: bash

                    nwb2bids convert ~/.nwb2bids/tutorials/ephys_tutorial_dataset/some_sessions ^
                        --bids-directory ~/.nwb2bids/tutorials/ephys_tutorial_dataset/some_sessions_bids_dataset

    .. tab:: Python Library

        .. code-block:: python

            >>> import pathlib
            >>>
            >>> import nwb2bids
            >>>
            >>> tutorial_directory = pathlib.Path.home() / ".nwb2bids/tutorials/ephys_tutorial_dataset"
            >>> nwb_paths = [tutorial_directory / "some_sessions"]
            >>> bids_directory = tutorial_directory / "some_sessions_bids_dataset"
            >>>
            >>> notifications = nwb2bids.convert_nwb_dataset(
            ...     nwb_paths=nwb_paths,
            ...     bids_directory=bids_directory,
            ... )

And our BIDS dataset should look like:

> .. code-block:: text

Reviewer: longer term:
- https://github.com/con/nwb2bids/issues/188

        ``--bids-directory``. These inputs can be any mix of files or directories.

    .. tab:: Python Library

        .. code-block:: python

            >>> import os
            >>> import pathlib
            >>>
            >>> import nwb2bids
            >>>
            >>> tutorial_directory = pathlib.Path.home() / ".nwb2bids/tutorials/ephys_tutorial_dataset"
            >>> nwb_paths = [
            ...     tutorial_directory / "ephys_session_3.nwb",
            ...     tutorial_directory / "some_sessions",
            ... ]
            >>> bids_directory = tutorial_directory / "implicit_bids_dataset"
            >>> bids_directory.mkdir(exist_ok=True)
            >>> os.chdir(path=bids_directory)
            >>>
            >>> notifications = nwb2bids.convert_nwb_dataset(nwb_paths=nwb_paths)

And the results should match what we saw at the end of :ref:`tutorial-multiple-inputs`.

.. _tutorial-additional-metadata:

Tutorial 5 - Additional metadata
---------------------------------

NWB files don't always include all the little metadata details you might want to include in BIDS. To include manual

> metadata that is unable to be extracted from the source files, you can provide additional metadata through a JSON

Reviewer: ```suggestion
metadata that is absent in the source files, you can provide additional metadata through a JSON
```

since otherwise reads odd, not sure if data could be able or unable ;)

structure that mirrors the BIDS specification nested under subject and session identifiers (which should match the

> ``subject_id`` and ``session_id`` found within each file).

Reviewer: this is unclear to me and no example below? moreover note says that it seems only this dataset_description is supported ATM.  Should we remove this for now?

Author: Yes, I will remove

Author: Done

To show how such additional metadata can be included through **nwb2bids**, start by creating a file named
``metadata.json`` inside the ``ephys_tutorial_dataset`` directory we used in :ref:`tutorial-multiple-files`.
Then fill the contents of the file to match below:

.. code-block:: json

    {
        "dataset_description": {
            "Name": "My Custom BIDS Dataset",
            "BIDSVersion": "1.8.0",

>             "Authors": ["First Last", "Second Author"]

Author: Oops, I should have known better...
